### PR TITLE
mysql/json: json_merge

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
    This PR implement `JSON_MERGE` for document, related PR in `tidb` is https://github.com/pingcap/tidb/pull/3374, and the relate source file in `tidb` is 
https://github.com/pingcap/tidb/blob/master/util/types/json/functions.go#L208

@hhkbp2 @hicqu @andelf @lishihai9017 PTAL

### CODE REVIEW

File: src/util/codec/mysql/json/json_merge.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        let mut data: Json = match self {
+            Json::Object(obj) => Json::Object(obj),
+            Json::Array(array) => Json::Array(array),
+            json => {
+                let mut array = Vec::with_capacity(suffixes.len() + 1);
+                array.push(json);
+                Json::Array(array)
+            }
+        };
+        let mut left_num = suffixes.len();
```
Comment (andelf) [inline src/util/codec/mysql/json/json_merge.rs]: normally lhs & rhs will be a good name.
`left_num` is ambiguous

File: src/util/codec/mysql/json/json_merge.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        for suffix in suffixes {
+            data = match (data, suffix) {
+                (Json::Array(mut array), Json::Array(mut sub_array)) => {
```
Comment (andelf) [inline src/util/codec/mysql/json/json_merge.rs]: will `ref mut` works?

File: src/util/codec/mysql/json/json_merge.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                (obj, Json::Array(mut sub_array)) => {
+                    // rule 4
+                    let mut array: Vec<Json> = Vec::with_capacity(sub_array.len() + left_num + 1);
+                    array.push(obj);
+                    array.append(&mut sub_array);
+                    Json::Array(array)
+                }
+                (obj, suffix) => {
+                    // rule 3, 4
+                    let mut array = Vec::with_capacity(left_num + 1);
```
Comment (andelf) [inline src/util/codec/mysql/json/json_merge.rs]: ? why left_num +1

File: src/util/codec/mysql/json/json_merge.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            Json::Object(obj) => Json::Object(obj),
+            Json::Array(array) => Json::Array(array),
+            json => {
+                let mut array = Vec::with_capacity(suffixes.len() + 1);
+                array.push(json);
+                Json::Array(array)
+            }
+        };
+        let mut left_num = suffixes.len();
+        for suffix in suffixes {
```
Comment (andelf) [inline src/util/codec/mysql/json/json_merge.rs]: this is a `Iterator::fold` pattern

File: src/util/codec/mysql/json/json_merge.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_merge() {
+        let test_cases = vec![
+            (r#"{"a": 1}"#,r#"{"b": 2}"#, r#"{"a": 1, "b": 2}"#),
```
Comment (andelf) [inline src/util/codec/mysql/json/json_merge.rs]: spaces

File: src/util/codec/mysql/json/json_merge.rs (hunk 2 above)
Comment (AndreMouche) [inline src/util/codec/mysql/json/json_merge.rs]: It seems not @andelf

**Comment by AndreMouche** [comment]
@lishihai9017 @andelf @hhkbp2 @hicqu PTAL

**Comment by hicqu** [comment]
LGTM.

**Comment by andelf** [comment]
LGTM

**Pull request merged**